### PR TITLE
Engine integrate server metrics

### PR DESCRIPTION
**Pull request by NahuFigueroa97**
|Related issue| Documentation| Testing issue|
|---|---|---|
| #16271| link to docu, user manual or design doc| Testing issue |

## Server metrics
Research and list Server metrics. Each metric must be clearly defined and documented, so its meaning and usage is clear. 
These metrics must help the administrator with capacity planning and debugging performance issues:

 - [x] Research and list Server metrics.
   - [x] Input socket connection in a point of time (Updown counter). Number of active connections.
   - [x] Number of events inserted in queue (Counter). 
   - [x] Number of events that have been written to disk because queue is full (Counter).
   - [x] Size in bytes received (Counter).
   - [x] Number of events received (Counter).
   - [x] Size in bytes per second received (Delta counter).
   - [x] Number of events per second received (Delta counter).
 - [x] Implement Server metrics.
 - [x] Update the user manual. 

Each module of the engine needs appropriate metrics as well. We must complete each module metrics after carefully considering its value.

### CODE REVIEW

File: src/engine/source/metrics/input_files/metrics-config.json [hunk 1]
```diff
@@ -1,76 +1,124 @@
...
+        "name": "Server.ReceivedBytesPerSeconds",
+        "dataHubEnable" : true,
+        "enable": true,
+        "signalType": "meter",
+        "instrumentType" : "counter",
+	    "aggregationTemporality" : "delta",
+        "subType": "uint64",
+	    "exportIntervalMillis": 1000,
+        "exportTimeoutMillis": 500,
+        "unit": "bytes/second",
```
Comment (soynof) [inline src/engine/source/metrics/input_files/metrics-config.json:12]: Verify the unit based on the official [documentation](https://opentelemetry-cpp.readthedocs.io/en/latest/).
We have to use the following [standard](https://unitsofmeasure.org/ucum).

File: src/engine/source/metrics/input_files/metrics-config.json [hunk 2]
```diff
@@ -1,76 +1,124 @@
...
+        "name": "Server.ReceivedEventsPerSeconds",
+        "dataHubEnable" : true,
+        "enable": true,
+        "signalType": "meter",
+        "instrumentType" : "counter",
+        "aggregationTemporality" : "delta",
+        "subType": "uint64",
+        "exportIntervalMillis": 1000,
+        "exportTimeoutMillis": 500,
+        "unit": "events/second",
```
Comment (soynof) [inline src/engine/source/metrics/input_files/metrics-config.json:25]: ditto

File: src/engine/source/metrics/input_files/metrics-config.json [hunk 3]
```diff
@@ -1,76 +1,124 @@
...
-        "name": "CountExample",
+        "name": "Server.TotalBytesReceived",
+        "dataHubEnable" : true,
+        "enable": true,
+        "signalType": "meter",
+        "instrumentType" : "counter",
+        "subType": "uint64",
+        "exportIntervalMillis": 100,
+        "exportTimeoutMillis": 50,
+        "unit": "bytes",
```
Comment (soynof) [inline src/engine/source/metrics/input_files/metrics-config.json:49]: ditto

File: src/engine/source/metrics/input_files/metrics-config.json [hunk 4]
```diff
@@ -1,76 +1,124 @@
...
         "enable": true,
         "signalType": "meter",
         "instrumentType" : "counter",
         "subType": "uint64",
-        "outputFile": "countExample.txt",
         "exportIntervalMillis": 100,
         "exportTimeoutMillis": 50,
-        "unit": "some unit",
-        "description": "some description"
+        "unit": "events",
```
Comment (soynof) [inline src/engine/source/metrics/input_files/metrics-config.json:61]: ditto

File: src/engine/source/metrics/input_files/metrics-config.json [hunk 5]
```diff
@@ -1,76 +1,124 @@
...
-        "instrumentType" : "histogram",
-        "subType": "double",
-        "outputFile": "histogramExample.txt",
+        "instrumentType" : "counter",
+        "subType": "uint64",
         "exportIntervalMillis": 100,
         "exportTimeoutMillis": 50,
-        "unit": "some unit",
-        "description": "some description"
+        "unit": "events",
```
Comment (soynof) [inline src/engine/source/metrics/input_files/metrics-config.json:73]: ditto

File: src/engine/source/metrics/input_files/metrics-config.json [hunk 6]
```diff
@@ -1,76 +1,124 @@
...
         "enable": true,
         "signalType": "meter",
         "instrumentType" : "upDownCounter",
         "subType": "int64",
-        "outputFile": "upDownCounterExample.txt",
         "exportIntervalMillis": 100,
         "exportTimeoutMillis": 50,
-        "unit": "some unit",
-        "description": "some description"
+        "unit": "connections",
```
Comment (soynof) [inline src/engine/source/metrics/input_files/metrics-config.json:85]: ditto

File: src/engine/source/metrics/input_files/metrics-config.json [hunk 7]
```diff
@@ -1,76 +1,124 @@
...
-        "outputFile": "tracerExample.txt",
-        "processorType": "simple",
-        "description": "some description"
+        "signalType": "meter",
+        "instrumentType" : "observableGauge",
+        "subType": "double",
+        "exportIntervalMillis": 1000,
+        "exportTimeoutMillis": 500,
+        "unit": "queue_occupancy_percentage",
+        "description": "This observableGauge was configured to measure Used queue"
```
Comment (soynof) [inline src/engine/source/metrics/input_files/metrics-config.json:110]: ditto

File: src/engine/source/metrics/input_files/metrics-config.json [hunk 8]
```diff
@@ -1,76 +1,124 @@
...
+    {
+        "name": "Metrics.TestCounter",
+        "dataHubEnable" : true,
+        "enable": true,
+        "signalType": "meter",
+        "instrumentType" : "counter",
+        "subType": "uint64",
+        "exportIntervalMillis": 100,
+        "exportTimeoutMillis": 50,
+        "unit": "unity",
```
Comment (soynof) [inline src/engine/source/metrics/input_files/metrics-config.json:121]: ditto

File: src/engine/source/metrics/input_files/test-metrics-config.json [hunk 9]
```diff
@@ -0,0 +1,54 @@
+[
```
Comment (soynof) [inline src/engine/source/metrics/input_files/test-metrics-config.json:1]: is this file necessary? what would it be used for?

File: src/engine/source/metrics/src/metrics.cpp [hunk 10]
```diff
@@ -98,6 +104,7 @@ void Metrics::createCommonChain(const std::filesystem::path& file)
 void Metrics::setMetricsConfig()
 {
     auto particularContext = m_upContext.begin();
+    m_instrumentState.clear();
```
Comment (soynof) [inline src/engine/source/metrics/src/metrics.cpp:107]: Why is it necessary to clean only this map?

File: src/engine/source/server/src/endpoints/baseEndpoint.hpp [hunk 11]
```diff
@@ -11,6 +11,7 @@
 #define _BASE_ENDPOINT_H
 
 #include <string>
+#include "metrics.hpp"
```
Comment (soynof) [inline src/engine/source/server/src/endpoints/baseEndpoint.hpp:14]: shouldn't this reference go in the .cpp file? and should we use <> instead of ""?

File: src/engine/test/source/metrics/otl_signals_test.cpp [hunk 12]
```diff
@@ -1,8 +1,9 @@
 #include "metrics.hpp"
 #include <gtest/gtest.h>
-#include <filesystem>
 
-const auto INPUT_PATH {(std::filesystem::current_path() / "source/metrics/input_files/metrics-config.json").string()};
+#include "metrics.hpp"
```
Comment (soynof) [inline src/engine/test/source/metrics/otl_signals_test.cpp:4]: Should we use <> instead of ""?

File: src/engine/test/source/metrics/otl_signals_test.cpp [hunk 13]
```diff
@@ -34,43 +35,37 @@ class MetricsOtlMeterTest : public ::testing::Test
 
 TEST_F(MetricsOtlMeterTest, invalidValueCounter)
 {
-    Metrics::instance().initMetrics("example", INPUT_PATH);
+    // TODO: Resolve fullpath accordingly.
+    Metrics::instance().initMetrics("engine-metrics", INPUT_PATH);
     EXPECT_ANY_THROW(Metrics::instance().addCounterValue("Sockets", -1UL));
 }
 
-TEST_F(MetricsOtlMeterTest, repeatedInit)
```
Comment (soynof) [inline src/engine/test/source/metrics/otl_signals_test.cpp:41]: Why removed?

File: src/engine/source/metrics/input_files/test-metrics-config.json (hunk 9 above)
Comment (NahuFigueroa97) [reply src/engine/source/metrics/input_files/test-metrics-config.json:1]: You have to separate the test metrics from those that are for the modules. Eventually this will disappear, but for now when you run the list command you only have to see the metrics of the modules.

File: src/engine/source/server/CMakeLists.txt [hunk 14]
```diff
@@ -54,4 +54,5 @@ target_include_directories(server
     ${RxCpp_SOURCE_DIR}/Rx/v2/src
     ${ENGINE_SERVER_SOURCE_DIR}/base
     ${ENGINE_SOURCE_DIR}/API/include
+    ${ENGINE_SOURCE_DIR}/cmds/src
```
Comment (JorgeMarinoDev) [inline src/engine/source/server/CMakeLists.txt:57]: Just to understand. Why do we need to include the /cmds/src to the target_include_directories ?

File: src/engine/source/metrics/src/metrics.cpp (hunk 10 above)
Comment (NahuFigueroa97) [reply src/engine/source/metrics/src/metrics.cpp:107]: this is because the map is initialized with default values so that the compiler does not complain about the empty constructor.

File: src/engine/source/server/src/endpoints/baseEndpoint.hpp (hunk 11 above)
Comment (JorgeMarinoDev) [reply src/engine/source/server/src/endpoints/baseEndpoint.hpp:14]: +1

File: src/engine/test/source/metrics/otl_signals_test.cpp (hunk 13 above)
Comment (NahuFigueroa97) [reply src/engine/test/source/metrics/otl_signals_test.cpp:41]: because to instantiate it you need the INPUT_PATH that should be defined in the main of the tests.

File: src/engine/test/source/metrics/otl_signals_test.cpp
Comment (JorgeMarinoDev) [inline src/engine/test/source/metrics/otl_signals_test.cpp:1]: I think we can keep the otl_signals_test.cpp changes outside the PR. Along with any other change outside the instrumented module.

File: src/engine/source/server/CMakeLists.txt (hunk 14 above)
Comment (NahuFigueroa97) [reply src/engine/source/server/CMakeLists.txt:57]: we need a file from that directory since it contains the maximum size of events in the queue. If you can't, it throws this error:
fatal error: defaultSettings.hpp: No such file or directory
    26 | #include "defaultSettings.hpp"
       | ^~~~~~~~~~~~~~~~~~~~~

File: src/engine/source/server/src/endpoints/eventEndpoint.cpp [hunk 15]
```diff
@@ -213,6 +215,19 @@ EventEndpoint::EventEndpoint(
...
+
+            // Received events per seconds
+            Metrics::instance().addCounterValue("Server.ReceivedEventsPerSeconds", 1UL);
+
+            // Total events received
+            Metrics::instance().addCounterValue("Server.TotalEventsReceived", 1UL);
+
+            // Total bytes received
+            Metrics::instance().addCounterValue("Server.TotalBytesReceived", static_cast<uint64_t>(eventSocket.length));
+
```
Comment (JorgeMarinoDev) [inline src/engine/source/server/src/endpoints/eventEndpoint.cpp:230]: Quick fix, typo: Seconds must be "second". i.e.   Events received per **second**.

**Review by soynof — APPROVED** [review]
Is OK!

**Review by JorgeMarinoDev — COMMENTED** [review]
Looks good to me. A few minor typos. Not sure to include changes in test or metrics module. These changes deserve their own PR and can also be tracked as an issue to record the effort put in collateral changes.

**Review by cborla — COMMENTED** [review]
lgtm
